### PR TITLE
Limit scope of libc usage in note-c.

### PR DESCRIPTION
**Pull request by haydenroche5**
See commit messages for more details.

### CODE REVIEW

File: n_cjson.c [hunk 1]
```diff
@@ -115,6 +121,21 @@ N_CJSON_PUBLIC(const char*) JVersion(void)
     return NOTE_C_STRINGIZE(N_CJSON_VERSION_MAJOR) "." NOTE_C_STRINGIZE(N_CJSON_VERSION_MINOR) "." NOTE_C_STRINGIZE(N_CJSON_VERSION_PATCH);
 }
 
+NOTE_C_STATIC char Jtolower(char c)
+{
+    // Less than A -> not a valid letter.
+    // Greater than z -> not a valid letter.
+    // Between a and Z -> not a valid letter.
+    // Between a and z (inclusive) -> already lower.
+    if (c < 'A' || c > 'z' || (c < 'a' && c > 'Z') || (c >= 'a' && c <= 'z')) {
```
Comment (zfields) [inline n_cjson.c:130]: This seems overly complicated.

It becomes more obvious if you swap in numbers.

Return the letter if:
x < 65 OR x > 122 OR (x < 97 AND x > 90) OR (x >= 97 AND x <= 122)

-128 >= x < 65,
x > 122,
90 < x < 97,
97 <= x <= 122

-128 <------ 65 ______ 91 --97--122--> 127

```suggestion
    if (c < 'A' || c > 'Z') {
```

File: scripts/check_libc_dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+PERMITTED_FNS=(
+    # The mem* functions are ok.
+    "memchr"
+    "memcmp"
+    "memcpy"
+    "memmove"
+    "memset"
+    # These string functions are ok.
+    "strchr"
+    "strcmp"
```
Comment (zfields) [inline scripts/check_libc_dependencies.sh:31]: This seems like a good opportunity to not allow these unsafe functions.

File: CMakeLists.txt [hunk 3]
```diff
@@ -18,6 +18,7 @@ option(NOTE_C_BUILD_TESTS "Build tests." OFF)
 option(NOTE_C_COVERAGE "Compile for test NOTE_C_COVERAGE reporting." OFF)
 option(NOTE_C_MEM_CHECK "Run tests with Valgrind." OFF)
 option(NOTE_C_BUILD_DOCS "Build docs." OFF)
+option(NOTE_C_NO_LIBC "Build the library without linking against libc, generating errors for any undefined symbols." OFF)
```
Comment (zfields) [inline CMakeLists.txt:21]: Is this supposed to be off, then gets switched on for specific test builds?

File: scripts/check_libc_dependencies.sh (hunk 2 above)
Comment (zfields) [reply scripts/check_libc_dependencies.sh:31]: FWIW, I don't think this should prevent this commit from happening, but we should add it to the backlog - especially now that there is a mechanism for protecting against this.

File: CMakeLists.txt (hunk 3 above)
Comment (haydenroche5) [reply CMakeLists.txt:21]: Yes.

File: n_cjson.c (hunk 1 above)
Comment (haydenroche5) [reply n_cjson.c:130]: Nah I think your change works.

File: scripts/check_libc_dependencies.sh (hunk 2 above)
Comment (haydenroche5) [reply scripts/check_libc_dependencies.sh:31]: I'd have to re-immerse myself in the discourse on this age-old argument, but my rough understanding is that you're not that much better off using strncmp over strcmp: https://stackoverflow.com/questions/24353504/whats-wrong-with-strcmp

If your string isn't NULL-terminated, all bets are off.

**Review by zfields — APPROVED** [review]

**Pull request merged**